output.c

```diff
@@ -64,8 +64,18 @@ void output_log_layers(struct liftoff_output *output) {
 	}
 
 	liftoff_log(LIFTOFF_DEBUG, "Layers on CRTC %"PRIu32":", output->crtc_id);
+
 	liftoff_list_for_each(layer, &output->layers, link) {
-		liftoff_log(LIFTOFF_DEBUG, "  Layer %p:", (void *)layer);
+		if (layer->force_composition) {
+			liftoff_log(LIFTOFF_DEBUG, "  Layer %p (forced composition):", (void *)layer);
```

Comment:
Since there's a `continue` statement right below, it's a little bit weird to have a colon at the end of the log. Can we remove the colon?

Reply:
I don't yet fully understand the "forced composition" mode. When composition is forced on a layer the FB_ID is 0 so what exactly is the layer for then, what is displayed on it?

Since the FB_ID is 0 how are the other props still relevant? If they are relevant in some way I guess we should drop the continue.

Reply:
Forcing composition is useful if you want to draw the layer's contents yourself. There are several use-cases:

- You need to draw a buffer coming from a Wayland client, but the client uses `wl_shm` so you can't get a DMA-BUF (thus can't get a GEM buffer ID)
- You want to apply a graphical effect to the buffer. For instance: blur, change the contrast, invert the colors.
- You want to draw an arbitrary shape with OpenGL (or whatever you're using for rendering): a solid circle, a gradient, and so on.

The layer's position and size is important when composition is forced, because layers underneath need to be composited as well.

It would be nice to document all of this in the header comments.

Reply:
But how do I draw anything on a layer without an FB_ID? Don't I need that in any case for associating some kind of pixel data with a layer/plane? Without a framebuffer to me a plane was always just an area of nothing.

> The layer's position and size is important when composition is forced, because layers underneath need to be composited as well.

Ok, so reading this it sounds like a layer with forced composition is kind of a virtual object to not directly be used by libliftoff in any way but to tell libliftoff about constraint for allocating planes to other "non-virtual" layers. Is this correct?

Then it could make sense to only print out the CRTC_* values for virtual layers.

Reply:
>But how do I draw anything on a layer without an FB_ID? Don't I need that in any case for associating some kind of pixel data with a layer/plane? Without a framebuffer to me a plane was always just an area of nothing.

The compositor won't be able to render anything with just an FB_ID. The compositor needs something like a GL texture ID to draw the buffer contents (typically obtained via importing a DMA-BUF with an EGL extension). For this reason, the compositor is supposed to keep some state alongside the `liftoff_layer` in case it needs to fallback to composition.

>Ok, so reading this it sounds like a layer with forced composition is kind of a virtual object to not directly be used by libliftoff in any way but to tell libliftoff about constraint for allocating planes to other "non-virtual" layers. Is this correct?

Yeah, that's correct. A layer which has composition forced will never make it into a plane. Its properties (CRTC_*, zpos) are used by libliftoff to add additional constraints (layers under it can't make it into a plane).

>Then it could make sense to only print out the CRTC_* values for virtual layers.

I think we could just print all properties, that would be simpler.

Reply:
> I think we could just print all properties, that would be simpler.

Yea, let's do that.